### PR TITLE
[WIP] Enable dotty

### PR DESCRIPTION
**Pull request by biboudis**
This is WIP to enable make the project dependent on both scalac and dotc. Also will need to make it compile as a simple scala 2.11.8, correct?

### CODE REVIEW

File: compilation/src/main/scala/scala/tools/nsc/WarmScalacBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+  * Created by bibou on 28/04/2017.
+  */
+@BenchmarkMode(Array(SampleTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 0)
+@Measurement(iterations = 1, time = 30, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 3)
+class WarmScalacBenchmark extends BaseScalacBenchmark {
```
Comment (olafurpg) [inline compilation/src/main/scala/scala/tools/nsc/WarmScalacBenchmark.scala]: Should we add Warm/Hot/ColdDotcBenchmarks?

File: compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  @Setup(Level.Trial)
+  def initTemp(): Unit = {
+    val tempFile = java.io.File.createTempFile("output", "")
+    tempFile.delete()
+    tempFile.mkdir()
+    tempDir = tempFile
+  }
+
+  @TearDown(Level.Trial)
+  def clearTemp(): Unit = {
```
Comment (olafurpg) [inline compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala]: Is this duplicated between dotc and scalac? If so, could we reuse the implementation?

Comment (olafurpg) [inline compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala]: Or what do you think about creating a `BaseCompilerBenchmark` that's shared between dotc and scalac?

**Comment by biboudis** [comment]
Still can't resolve the dependency correctly of dotc, that's the reason
behind the WIP :( :(.

I'll try to finish this PR during the WE. I'm thinking about removing as
much as I can from the actual benchmark method and maybe having a common
base class at the end.

WDYT about the dependency resolution? Any quick feedback? Do you think it
may be better to have two separate project definitions?



On Apr 28, 2017 4:13 PM, "Ólafur Páll Geirsson" <notifications@github.com>
wrote:

*@olafurpg* commented on this pull request.

Awesome! Thanks for the quick setup @biboudis <https://github.com/biboudis>

Did you manage to run the benchmarks? I tried to compile locally and got

[info] Compiling 5 Scala sources and 2 Java sources to
/Users/ollie/dev/compiler-benchmark/compilation/target/scala-2.11/classes...
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala:22:
unbound placeholder parameter
[error]   var compilerArgs = _
[error]                      ^
[error] one error found
[error] (compilation/compile:compileIncremental) Compilation failed
[error] Total time: 3 s, completed Apr 28, 2017 4:10:17 PM
[enable-dotty] compiler-benchmark: hot -psource=vector -i 1 -f 1 -wi 0
-rff target/foo.csv
[info] Compiling 5 Scala sources and 2 Java sources to
/Users/ollie/dev/compiler-benchmark/compilation/target/scala-2.11/classes...
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseDotcBenchmark.scala:43:
not found: value tempOutDir
[error]     ctx.setSetting(ctx.settings.d, tempOutDir.getAbsolutePath)
[error]                                    ^
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseDotcBenchmark.scala:44:
value nowarn is not a member of dotty.tools.dotc.config.ScalaSettings
[error]     ctx.setSetting(ctx.settings.nowarn, true)
[error]                                 ^
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseDotcBenchmark.scala:47:
not found: value Bench
[error]     val reporter = Bench.doCompile(new Compiler, compilerArgs.toList)
[error]                    ^
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseDotcBenchmark.scala:47:
Compiler does not have a constructor
[error]     val reporter = Bench.doCompile(new Compiler, compilerArgs.toList)
[error]                                    ^
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala:45:
type mismatch;
[error]  found   : Array[String]
[error]  required: String
[error]         Array(source)
[error]              ^
[error] /Users/ollie/dev/compiler-benchmark/compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala:56:
type mismatch;
[error]  found   : String
[error]  required: Array[String]
[error]     driver.process(compilerArgs)
[error]                    ^
[error] 6 errors found
[error] (compilation/compile:compileIncremental) Compilation failed

------------------------------

In compilation/src/main/scala/scala/tools/nsc/WarmScalacBenchmark.scala
<https://github.com/olafurpg/compiler-benchmark/pull/1#discussion_r113938010>
:

> +package scala.tools.nsc
+
+import java.util.concurrent.TimeUnit
+
+import org.openjdk.jmh.annotations.Mode.SampleTime
+import org.openjdk.jmh.annotations._
+
+/**
+  * Created by bibou on 28/04/2017.
+  */
+@BenchmarkMode(Array(SampleTime))
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 0)
+@Measurement(iterations = 1, time = 30, timeUnit = TimeUnit.SECONDS)
+@Fork(value = 3)
+class WarmScalacBenchmark extends BaseScalacBenchmark {

Should we add Warm/Hot/ColdDotcBenchmarks?
------------------------------

In compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala
<https://github.com/olafurpg/compiler-benchmark/pull/1#discussion_r113938242>
:

> +    driver.process(compilerArgs)
+    assert(!driver.reporter.hasErrors) // TODO: Remove
+  }
+
+  private var tempDir: File = null
+
+  @Setup(Level.Trial)
+  def initTemp(): Unit = {
+    val tempFile = java.io.File.createTempFile("output", "")
+    tempFile.delete()
+    tempFile.mkdir()
+    tempDir = tempFile
+  }
+
+  @TearDown(Level.Trial)
+  def clearTemp(): Unit = {

Is this duplicated between dotc and scalac? If so, could we reuse the
implementation?
------------------------------

In compilation/src/main/scala/scala/tools/nsc/BaseScalacBenchmark.scala
<https://github.com/olafurpg/compiler-benchmark/pull/1#discussion_r113938409>
:

> +    driver.process(compilerArgs)
+    assert(!driver.reporter.hasErrors) // TODO: Remove
+  }
+
+  private var tempDir: File = null
+
+  @Setup(Level.Trial)
+  def initTemp(): Unit = {
+    val tempFile = java.io.File.createTempFile("output", "")
+    tempFile.delete()
+    tempFile.mkdir()
+    tempDir = tempFile
+  }
+
+  @TearDown(Level.Trial)
+  def clearTemp(): Unit = {

Or what do you think about creating a BaseCompilerBenchmark that's shared
between dotc and scalac?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/olafurpg/compiler-benchmark/pull/1#pullrequestreview-35369129>,
or mute the thread

**Comment by olafurpg** [comment]
Have you tried ?

```
export COMPILERVERSION=0.1.1-bin-20170427-fa13f8d-NIGHTLY
```

That should be picked up by sbt-dotty

**Comment by olafurpg** [comment]
Or just hardcode `dottyVersion := "..."` to that version number, that should be good for now. sbt-dotty will change soon anyways

**Comment by biboudis** [comment]
Can you scalafix this @olafurpg? (I am starting to use scalafix as a verb! ;-D)

```
[error] -- [E039] Syntax Error: /Users/bibou/Projects/EPFL/compiler-benchmark/compilation/../corpus/vector/Vector.scala:89:75 
[error] 89 |  private[collection] final def initIterator[B >: A](s: VectorIterator[B]) {
[error]    |                                                                           ^
[error]    |                                             '=' expected, but `B` found
[error] 
[error] longer explanation available when compiling with `-explain`
[error] 
[error] one error found
```

**Comment by biboudis** [comment]
Hm. Getting an IncompatibleClassChangeError

```
[info] java.lang.IncompatibleClassChangeError: Implementing class
[info]  at java.lang.ClassLoader.defineClass1(Native Method)
[info]  at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
[info]  at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[info]  at java.net.URLClassLoader.defineClass(URLClassLoader.java:467)
[info]  at java.net.URLClassLoader.access$100(URLClassLoader.java:73)
[info]  at java.net.URLClassLoader$1.run(URLClassLoader.java:368)
[info]  at java.net.URLClassLoader$1.run(URLClassLoader.java:362)
[info]  at java.security.AccessController.doPrivileged(Native Method)
[info]  at java.net.URLClassLoader.findClass(URLClassLoader.java:361)
[info]  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
[info]  at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:331)
[info]  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
[info]  at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:53)
[info]  at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:283)
[info]  at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:281)
[info]  at scala.collection.immutable.List.map(List.scala:273)
[info]  at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:281)
[info]  at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:38)
[info]  at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:76)
[info]  at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:73)
[info]  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
[info]  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
[info]  at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:73)
[info]  at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
[info]  at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
[info]  at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
[info]  at dotty.tools.dotc.Run.compileUnits(Run.scala:67)
[info]  at dotty.tools.dotc.Run.compileSources(Run.scala:64)
[info]  at dotty.tools.dotc.Run.compile(Run.scala:48)
[info]  at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
[info]  at dotty.tools.dotc.Bench$.dotty$tools$dotc$Bench$$super$doCompile(Bench.scala:27)
[info]  at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:27)
[info]  at dotty.tools.dotc.Bench$$anonfun$doCompile$1.apply(Bench.scala:25)
[info]  at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:22)
[info]  at dotty.tools.dotc.Bench$$anonfun$ntimes$1.apply(Bench.scala:22)
[info]  at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
[info]  at scala.collection.TraversableOnce$$anonfun$foldLeft$1.apply(TraversableOnce.scala:157)
[info]  at scala.collection.immutable.Range.foreach(Range.scala:160)
[info]  at scala.collection.TraversableOnce$class.foldLeft(TraversableOnce.scala:157)
[info]  at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
[info]  at scala.collection.TraversableOnce$class.$div$colon(TraversableOnce.scala:151)
[info]  at scala.collection.AbstractTraversable.$div$colon(Traversable.scala:104)
[info]  at dotty.tools.dotc.Bench$.ntimes(Bench.scala:22)
[info]  at dotty.tools.dotc.Bench$.doCompile(Bench.scala:25)
[info]  at base.BaseBenchmark.compileDotc(BaseBenchmark.scala:94)
[info]  at dotty.tools.dotc.HotDotcBenchmark.compileDotc(HotScalacBenchmark.scala:16)
[info]  at dotty.tools.dotc.generated.HotDotcBenchmark_compileDotc_jmhTest.compileDotc_sample_jmhStub(HotDotcBenchmark_compileDotc_jmhTest.java:309)
[info]  at dotty.tools.dotc.generated.HotDotcBenchmark_compileDotc_jmhTest.compileDotc_SampleTime(HotDotcBenchmark_compileDotc_jmhTest.java:247)
[info]  at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[info]  at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[info]  at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[info]  at java.lang.reflect.Method.invoke(Method.java:497)
[info]  at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:444)
[info]  at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:426)
[info]  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]  at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]  at java.lang.Thread.run(Thread.java:745)

```

**Comment by biboudis** [comment]
Can be reproduced just with ```compilation/jmh:run HotDotcBenchmark -p source=test```.

**Comment by jvican** [comment]
My port of the compiler benchmarks was working and revealed some issues with scalap. What's the value of porting the dotty benchmarks again here @biboudis ?
  
That error does not ring a bell, but it sounds like an issue with the classpath and the format of the class files. Do you have Scalac-generated class files in the classpath?

**Comment by olafurpg** [comment]
@jvican we used your port and rebased it on master. The additional change we made was to keep the scalac benchmarks for comparison with dotc. The incompatible error appears to come from scala-compiler and dotty-compiler being on the same classpath, commenting out scala-compiler from build.sbt fixes the problem and the benchmarks run.

@biboudis here are the results from running on my macbook


```
[info] Benchmark                                         (classPath)  (dottyVersion)  (extraArgs)  (source)    Mode  Cnt    Score   Error  Units
[info] HotDotcBenchmark.compileDotc                                                                  vector  sample  965  315.169 ± 2.526  ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.00                                                vector  sample       289.407          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.50                                                vector  sample       309.330          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.90                                                vector  sample       336.593          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.95                                                vector  sample       349.700          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.99                                                vector  sample       404.593          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.999                                               vector  sample       605.028          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p0.9999                                              vector  sample       605.028          ms/op
[info] HotDotcBenchmark.compileDotc:compileDotc·p1.00                                                vector  sample       605.028          ms/op



[info] Benchmark                                               (classPath)  (dottyVersion)  (extraArgs)  (source)    Mode  Cnt    Score   Error  Units
[info] HotScalacBenchmark.compileScalac                                                                    vector  sample  794  383.137 ± 3.126  ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.00                                                vector  sample       345.506          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.50                                                vector  sample       377.487          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.90                                                vector  sample       412.877          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.95                                                vector  sample       425.329          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.99                                                vector  sample       455.108          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.999                                               vector  sample       737.149          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p0.9999                                              vector  sample       737.149          ms/op
[info] HotScalacBenchmark.compileScalac:compileScalac·p1.00                                                vector  sample       737.149          ms/op
``` 

This is at least a start. Judging from the numbers, scalac 383ms vs. dotc 315ms, hot scalac is 20% slower compiling vector.scala than a hot dotc. Is that a reasonable interpretation? 

Running both took ~30 minutes on my laptop, and that's only the hot benchmark. We definitely want to automate this for warm/cold/hot and also more sources than vector.

**Comment by olafurpg** [comment]
Here are the commit I ran 

dotc https://github.com/olafurpg/compiler-benchmark/commit/785f13b18b900e2d78d85f427b013f280fde91ea
scalac https://github.com/olafurpg/compiler-benchmark/commit/6b57d43a5dd15839fda481067da98fd02b117193

**Comment by jvican** [comment]
> This is at least a start. Judging from the numbers, scalac 383ms vs. dotc 315ms, hot scalac is 20% slower compiling vector.scala than a hot dotc. Is that a reasonable interpretation?
  
Yes, though that's the average, I would use percentiles instead:

* https://developer.nvidia.com/content/analysing-stutter-–-mining-more-percentiles-0
* https://www.azul.com/files/HowNotToMeasureLatency_LLSummit_NYC_12Nov2013.

**Comment by biboudis** [comment]
Thanks @olafurpg, @jvican, that's a good start, I agree. We need to modularize this. Maybe split the compilation project into two would be best. I am putting the numbers in the google spreadsheet just for fun! :-)

**Pull request merged**